README.md
<!-- Please use absolute URLs for all links so that they work when the content of this page is rendered in docs.pact.io -->

# Pact Go

Golang version of [Pact](http://pact.io). Pact is a contract testing framework for HTTP APIs and non-HTTP asynchronous messaging systems.

Enables consumer driven contract testing, providing a mock service and
DSL for the consumer project, and interaction playback and verification for the service Provider project.

[![Build Status](https://travis-ci.org/JEFFTheDev/pact-go.svg?branch=master)](https://travis-ci.org/JEFFTheDev/pact-go)
[![Build status](https://ci.appveyor.com/api/projects/status/lg02mfcmvr3e8w5n?svg=true)](https://ci.appveyor.com/project/mefellows/pact-go)
[![Coverage Status](https://coveralls.io/repos/github/JEFFTheDev/pact-go/badge.svg?branch=HEAD)](https://coveralls.io/github/JEFFTheDev/pact-go?branch=HEAD)
[![Go Report Card](https://goreportcard.com/badge/github.com/JEFFTheDev/pact-go)](https://goreportcard.com/report/github.com/JEFFTheDev/pact-go)
[![GoDoc](https://godoc.org/github.com/JEFFTheDev/pact-go?status.svg)](https://godoc.org/github.com/JEFFTheDev/pact-go)
[![slack](https://slack.pact.io/badge.svg)](https://slack.pact.io)
[![Known Vulnerabilities](https://snyk.io/test/github/JEFFTheDev/pact-go/badge.svg?targetFile=Gopkg.lock)](https://snyk.io/test/github/JEFFTheDev/pact-go?targetFile=Gopkg.lock)

## Introduction

From the [Pact website]:

> The Pact family of frameworks provide support for [Consumer Driven Contracts](http://martinfowler.com/articles/consumerDrivenContracts.html) testing.

> A Contract is a collection of agreements between a client (Consumer) and an API (Provider) that describes the interactions that can take place between them.

> Consumer Driven Contracts is a pattern that drives the development of the Provider from its Consumers point of view.

> Pact is a testing tool that guarantees those Contracts are satisfied.

Read [Getting started with Pact] for more information for beginners.

<p align="center">
  <a href="https://asciinema.org/a/180671">
    <img width="880" src="https://raw.githubusercontent.com/JEFFTheDev/pact-go/master/.github/pact-small.svg?sanitize=true"></img>
  </a>
</p>

## Table of Contents

<!-- TOC -->

- [Pact Go](#pact-go)
  - [Introduction](#introduction)
  - [Table of Contents](#table-of-contents)
  - [Versions](#versions)
  - [Installation](#installation)
    - [Go get](#go-get)
    - [Installation on \*nix](#installation-on-\nix)
  - [V3 Beta](#v3-beta)
  - [Using Pact](#using-pact)
  - [HTTP API Testing](#http-api-testing)
    - [Consumer Side Testing](#consumer-side-testing)
    - [Provider API Testing](#provider-api-testing)
      - [Provider Verification](#provider-verification)
      - [Provider States](#provider-states)
      - [Before and After Hooks](#before-and-after-hooks)
      - [Request Filtering](#request-filtering)
        - [Example: API with Authorization](#example-api-with-authorization)
      - [Pending Pacts](#pending-pacts)
      - [WIP Pacts](#wip-pacts)
      - [Lifecycle of a provider verification](#lifecycle-of-a-provider-verification)
    - [Publishing pacts to a Pact Broker and Tagging Pacts](#publishing-pacts-to-a-pact-broker-and-tagging-pacts)
      - [Publishing from Go code](#publishing-from-go-code)
      - [Publishing Provider Verification Results to a Pact Broker](#publishing-provider-verification-results-to-a-pact-broker)
      - [Publishing from the CLI](#publishing-from-the-cli)
      - [Using the Pact Broker with Basic authentication](#using-the-pact-broker-with-basic-authentication)
      - [Using the Pact Broker with Bearer Token authentication](#using-the-pact-broker-with-bearer-token-authentication)
  - [Asynchronous API Testing](#asynchronous-api-testing)
    - [Consumer](#consumer)
    - [Provider (Producer)](#provider-producer)
    - [Pact Broker Integration](#pact-broker-integration)
  - [Matching](#matching)
    - [Matching on types](#matching-on-types)
    - [Matching on arrays](#matching-on-arrays)
    - [Matching by regular expression](#matching-by-regular-expression)
    - [Match common formats](#match-common-formats)
      - [Auto-generate matchers from struct tags](#auto-generate-matchers-from-struct-tags)
  - [Tutorial (60 minutes)](#tutorial-60-minutes)
  - [Examples](#examples)
    - [HTTP APIs](#http-apis)
    - [Asynchronous APIs](#asynchronous-apis)
    - [Integrated examples](#integrated-examples)
  - [Troubleshooting](#troubleshooting)
      - [Splitting tests across multiple files](#splitting-tests-across-multiple-files)
      - [Output Logging](#output-logging)
      - [Check if the CLI tools are up to date](#check-if-the-cli-tools-are-up-to-date)
      - [Disable CLI checks to speed up tests](#disable-cli-checks-to-speed-up-tests)
      - [Re-run a specific provider verification test](#re-run-a-specific-provider-verification-test)
    - [Verifying APIs with a self-signed certificate](#verifying-apis-with-a-self-signed-certificate)
    - [Testing AWS API Gateway APIs](#testing-aws-api-gateway-apis)
  - [Contact](#contact)
  - [Documentation](#documentation)
  - [Roadmap](#roadmap)
  - [Contributing](#contributing)

<!-- /TOC -->
## Versions

<details><summary>Specification Compatibility</summary>

| Version | Stable | [Spec] Compatibility | Install            |
|---------|--------|----------------------|--------------------|
| 2.x.x   | Beta   | 2, 3                 | See [v2.x.x] |
| 1.0.x   | Yes    | 2, 3\*               | See [installation] |
| 0.x.x   | Yes    | Up to v2             | 0.x.x [stable]     |

_\*_ v3 support is limited to the subset of functionality required to enable language inter-operable [Message support].

</details>

## Installation

1.  Download the latest [CLI tools] of the standalone tools and ensure the binaries are on your `PATH`:
1.  Unzip the package into a known location, and ensuring the `pact` and other binaries in the `bin` directory are on the `PATH`.
1.  Run `go get github.com/JEFFTheDev/pact-go@v1` to install the source packages

See below for how to automate this:

### Go get

Since `1.x.x` Pact is go-gettable, and uses tags for versioning, so `dep ensure --add github.com/JEFFTheDev/pact-go@1.0.0` or `go get gopkg.in/JEFFTheDev/pact-go.v1` is now possible.

See the [Changelog] for versions to pin to and their history.

### Installation on \*nix

The following will install pact binaries into `/opt/pact/bin`:

```sh
cd /opt
curl -fsSL https://raw.githubusercontent.com/JEFFTheDev/pact-ruby-standalone/master/install.sh | bash
export PATH=$PATH:/opt/pact/bin
go get github.com/JEFFTheDev/pact-go@v1
```

Test the installation:

```sh
pact help
```

_NOTE_: the above script installs the latest standalone tools at the time it was ran. It is recommended you pin the installation to a [specific version](https://github.com/JEFFTheDev/pact-ruby-standalone/releases) of a release so that you may control the upgrade cycle.

## V3 Beta

If you are interested in testing out the new new beta package that supports all of the V3 [spec], and moves to a rust shared core, please head to [v2.x.x] and also let us know on [slack].

## Using Pact

Pact supports [synchronous request-response style HTTP interactions](#http-api-testing) and has experimental support for [asynchronous interactions](#asynchronous-api-testing) with JSON-formatted payloads.

Pact Go runs as part of your regular Go tests.

## HTTP API Testing

### Consumer Side Testing

We'll run through a simple example to get an understanding the concepts:

1.  `go get github.com/JEFFTheDev/pact-go`
1.  `cd $GOPATH/src/github.com/JEFFTheDev/pact-go/examples/`
1.  `go test -v -run TestConsumer`.

The simple example looks like this:

```go
func TestConsumer(t *testing.T) {
	type User struct {
		Name     string `json:"name" pact:"example=billy"`
		LastName string `json:"lastName" pact:"example=sampson"`
	}

	// Create Pact connecting to local Daemon
	pact := &dsl.Pact{
		Consumer: "MyConsumer",
		Provider: "MyProvider",
		Host:     "localhost",
	}
	defer pact.Teardown()

	// Pass in test case. This is the component that makes the external HTTP call
	var test = func() (err error) {
		u := fmt.Sprintf("http://localhost:%d/foobar", pact.Server.Port)
		req, err := http.NewRequest("GET", u, strings.NewReader(`{"name":"billy"}`))
		if err != nil {
			return
		}

		// NOTE: by default, request bodies are expected to be sent with a Content-Type
		// of application/json. If you don't explicitly set the content-type, you
		// will get a mismatch during Verification.
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Authorization", "Bearer 1234")

		_, err = http.DefaultClient.Do(req)
		return
	}

	// Set up our expected interactions.
	pact.
		AddInteraction().
		Given("User foo exists").
		UponReceiving("A request to get foo").
		WithRequest(dsl.Request{
			Method:  "GET",
			Path:    dsl.String("/foobar"),
			Headers: dsl.MapMatcher{"Content-Type": dsl.String("application/json"), "Authorization": dsl.String("Bearer 1234")},
			Body: map[string]string{
				"name": "billy",
			},
		}).
		WillRespondWith(dsl.Response{
			Status:  200,
			Headers: dsl.MapMatcher{"Content-Type": dsl.String("application/json")},
			Body:    dsl.Match(&User{}),
		})

	// Run the test, verify it did what we expected and capture the contract
	if err := pact.Verify(test); err != nil {
		log.Fatalf("Error on Verify: %v", err)
	}

	return nil
}

```

### Provider API Testing

1.  `go get github.com/JEFFTheDev/pact-go`
1.  `cd $GOPATH/src/github.com/JEFFTheDev/pact-go/examples/`
1.  `go test -v -run TestProvider`.

Here is the Provider test process broker down:

1.  Start your Provider API:

    You need to be able to first start your API in the background as part of your tests
    before you can run the verification process. Here we create `startServer` which can be
    started in its own goroutine:

    ```go
    var lastName = "" // User doesn't exist
    func startServer() {
      mux := http.NewServeMux()

      mux.HandleFunc("/users", func(w http.ResponseWriter, req *http.Request) {
        w.Header().Add("Content-Type", "application/json")
        fmt.Fprintf(w, fmt.Sprintf(`{"lastName":"%s"}`, lastName))
      })
      log.Fatal(http.ListenAndServe(":8000", mux))
    }
    ```

2.  Verify provider API

    You can now tell Pact to read in your Pact files and verify that your API will
    satisfy the requirements of each of your known consumers:

    ```go
    func TestProvider(t *testing.T) {

      // Create Pact connecting to local Daemon
      pact := &dsl.Pact{
        Provider: "MyProvider",
      }

      // Start provider API in the background
      go startServer()

      // Verify the Provider using the locally saved Pact Files
      pact.VerifyProvider(t, types.VerifyRequest{
        ProviderBaseURL:        "http://localhost:8000",
        PactURLs:               []string{filepath.ToSlash(fmt.Sprintf("%s/myconsumer-myprovider.json", pactDir))},
        StateHandlers: types.StateHandlers{
          // Setup any state required by the test
          // in this case, we ensure there is a "user" in the system
          "User foo exists": func() error {
            lastName = "crickets"
            return nil
          },
        },
      })
    }
    ```

The `VerifyProvider` will handle all verifications, treating them as subtests
and giving you granular test reporting. If you don't like this behaviour, you may call `VerifyProviderRaw` directly and handle the errors manually.

Note that `PactURLs` may be a list of local pact files or remote based
urls (e.g. from a
[Pact Broker](https://docs.pact.io/pact_broker)).

#### Provider Verification

When validating a Provider, you have 3 options to provide the Pact files:

1.  Use `PactURLs` to specify the exact set of pacts to be replayed:

    ```go
    pact.VerifyProvider(t, types.VerifyRequest{
    	ProviderBaseURL:        "http://myproviderhost",
    	PactURLs:               []string{"http://broker/pacts/provider/them/consumer/me/latest/dev"},
    	BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
    	BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
    	BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
    })
    ```

1.  Use `BrokerURL` to automatically find all of the latest consumers:

    ```go
    pact.VerifyProvider(t, types.VerifyRequest{
    	ProviderBaseURL:        "http://myproviderhost",
    	BrokerURL:              "http://brokerHost",
    	BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
    	BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
    	BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
    })
    ```

1.  Use `BrokerURL` and `Tags` to automatically find all of the latest consumers given one or more tags:

    ```go
    pact.VerifyProvider(t, types.VerifyRequest{
    	ProviderBaseURL:        "http://myproviderhost",
    	BrokerURL:              "http://brokerHost",
    	Tags:                   []string{"master", "prod"},
    	BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
    	BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
    	BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
    })
    ```

Options 2 and 3 are particularly useful when you want to validate that your
Provider is able to meet the contracts of what's in Production and also the latest
in development.

See this [article](http://rea.tech/enter-the-pact-matrix-or-how-to-decouple-the-release-cycles-of-your-microservices/)
for more on this strategy.

#### Provider States

If you have defined any states (as denoted by a `Given()`) in your consumer tests, the `Verifier` can put the provider into the correct state prior to sending the actual request for validation. For example, the provider can use the state to mock away certain database queries. To support this, set up a `StateHandler` for each state using hooks on the `StateHandlers` property. Here is an example:

```go
pact.VerifyProvider(t, types.VerifyRequest{
	...
  StateHandlers: types.StateHandlers{
		"User jmarie exists": func() error {
			userRepository = jmarieExists
			return nil
		},
		"User jmarie is unauthenticated": func() error {
			userRepository = jmarieUnauthorized
			token = "invalid"

			return nil
		},
		"User jmarie does not exist": func() error {
			userRepository = jmarieDoesNotExist
			return nil
		},
		...
	},
})
```

As you can see, for each state (`"User jmarie exists"` etc.) we configure the local datastore differently. If this option is not configured, the `Verifier` will ignore the provider states defined in the pact and log a warning.

Note that if the State Handler errors, the test will exit early with a failure.

Read more about [Provider States](https://docs.pact.io/getting_started/provider_states).

#### Before and After Hooks

Sometimes, it's useful to be able to do things before or after a test has run, such as reset a database, log a metric etc. A `BeforeEach` runs before any other part of the Pact test lifecycle, and a `AfterEach` runs as the last step before returning the verification result back to the test.

You can add them to your Verification as follows:

```go
	pact.VerifyProvider(t, types.VerifyRequest{
		...
		BeforeEach: func() error {
			fmt.Println("before hook, do something")
			return nil
		},
		AfterEach: func() error {
			fmt.Println("after hook, do something")
			return nil
		},
	})
```

If the Hook errors, the test will fail.

#### Request Filtering

Sometimes you may need to add things to the requests that can't be persisted in a pact file. Examples of these are authentication tokens with a small life span. e.g. an OAuth bearer token: `Authorization: Bearer 0b79bab50daca910b000d4f1a2b675d604257e42`.

For these cases, we have two facilities that should be carefully used during verification:

1. the ability to specify custom headers to be sent during provider verification. The flag to achieve this is `CustomProviderHeaders`.
1. the ability to modify a request/response and change the payload. The parameter to achieve this is `RequestFilter`.

Read on for more.

##### Example: API with Authorization

**Custom Headers**:

This header will always be sent for each and every request, and can't be dynamic. For example:

```go
  pact.VerifyProvider(t, types.VerifyRequest{
    ...
    CustomProviderHeaders:  []string{"Authorization: Bearer 0b79bab50daca910b000d4f1a2b675d604257e42"},
  })
```

As you can see, this is your opportunity to modify\add to headers being sent to the Provider API, for example to create a valid time-bound token.

**Request Filters**

_WARNING_: This should only be attempted once you know what you're doing!

Request filters are custom middleware, that are executed for each request, allowing `token` to change between invocations. Request filters can change the request coming in, _and_ the response back to the verifier. It is common to pair this with `StateHandlers` as per above, that can set/expire the token
for different test cases:

```go
  pact.VerifyProvider(t, types.VerifyRequest{
    ...
    RequestFilter: func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			r.Header.Add("Authorization", fmt.Sprintf("Bearer %s", token))
			next.ServeHTTP(w, r)
		})
	}
  })
```

_Important Note_: You should only use this feature for things that can not be persisted in the pact file. By modifying the request, you are potentially modifying the contract from the consumer tests!

#### Pending Pacts
_NOTE_: This feature is currently only available on [Pactflow]

Pending pacts is a feature that allows consumers to publish new contracts or changes to existing contracts without breaking Provider's builds. It does so by flagging the contract as "unverified" in the Pact Broker the first time a contract is published. A Provider can then enable a behaviour (via `EnablePending: true`) that will still perform a verification (and thus share the results back to the broker) but _not_ fail the verification step itself.

This enables safe introduction of new contracts into the system, without breaking Provider builds, whilst still providing feedback to Consumers as per before.

See the [docs](https://docs.pact.io/pending) and this [article](http://blog.pact.io/2020/02/24/how-we-have-fixed-the-biggest-problem-with-the-pact-workflow/) for more background.

#### WIP Pacts

_NOTE_: This feature is currently only available on [Pactflow]


CHANGELOG.md
# Changelog
Do this to generate your change history

    git log --pretty=format:'  * [%h](https://github.com/JEFFTheDev/pact-go/commit/%h) - %s (%an, %ad)' vX.Y.Z..HEAD | egrep -v "wip(:|\()" | grep -v "docs(" | grep -v "chore(" | grep -v Merge | grep -v "test("

## Versions

### v1.6.9 (09 February 2022)
  * [25a5ad5](https://github.com/JEFFTheDev/pact-go/commit/25a5ad5) - Match should skip unexported fields as they won't be serialized (#201) (Cody A. Ray, Tue Feb 8 21:07:54 2022 -0600)

### v1.6.8 (04 February 2022)
  * [18c791e](https://github.com/JEFFTheDev/pact-go/commit/18c791e) - fix: support struct tags for pointers (Sonia Hamilton, Tue Feb 1 10:31:07 2022 +1100)
  * [0360e55](https://github.com/JEFFTheDev/pact-go/commit/0360e55) - refactor(all): Small refactor (#193) (Sasha Melentyev, Sun Dec 26 15:44:11 2021 +0300)
  * [c8a5f24](https://github.com/JEFFTheDev/pact-go/commit/c8a5f24) - refactor(all): small refactor (#192) (Sasha Melentyev, Fri Dec 24 05:40:37 2021 +0300)

### v1.6.7 (20 December 2021)
  * [5160a01](https://github.com/JEFFTheDev/pact-go/commit/5160a01) - fix: allow provider log level to be set/logged to file (Matt Fellows, Mon Dec 20 16:28:15 2021 +1100)

### v1.6.6 (17 November 2021)
  * [ad35b25](https://github.com/JEFFTheDev/pact-go/commit/ad35b25) - feat: support matching branch selector. Fixes #184 (Matt Fellows, Wed Nov 17 17:32:44 2021 +1100)

### v1.6.5 (09 November 2021)
  * [030da88](https://github.com/JEFFTheDev/pact-go/commit/030da88) - chore: go mod tidy (David Fleming, Mon Nov 8 12:17:25 2021 -0500)
  * [3dc7736](https://github.com/JEFFTheDev/pact-go/commit/3dc7736) - feat: update to standalone 1.88.78 and support new verifier option: --provider-version-branch (David Fleming, Thu Nov 4 15:47:19 2021 -0400)
  * [511f144](https://github.com/JEFFTheDev/pact-go/commit/511f144) - fix: Dockerfile to reduce vulnerabilities (snyk-bot, Thu Oct 14 07:22:29 2021 +0000)
  * [493f853](https://github.com/JEFFTheDev/pact-go/commit/493f853) - fix: Dockerfile to reduce vulnerabilities (snyk-bot, Thu Sep 30 05:23:55 2021 +0000)
  * [7549f57](https://github.com/JEFFTheDev/pact-go/commit/7549f57) - chore: update installation docs to recommend pinning standalone package (Matt Fellows, Thu Sep 9 13:53:27 2021 +1000)
  * [6ab2895](https://github.com/JEFFTheDev/pact-go/commit/6ab2895) - build(deps): bump github.com/gin-gonic/gin from 1.6.3 to 1.7.2 (dependabot-preview[bot], Mon May 24 19:23:29 2021 +0000)
  * [fccd0c7](https://github.com/JEFFTheDev/pact-go/commit/fccd0c7) - Upgrade to GitHub-native Dependabot (dependabot-preview[bot], Wed Apr 28 22:24:10 2021 +0000)
  * [942b1a2](https://github.com/JEFFTheDev/pact-go/commit/942b1a2) - build(deps): bump github.com/hashicorp/go-version from 1.0.0 to 1.3.0 (dependabot-preview[bot], Wed Mar 31 19:25:44 2021 +0000)
  * [0975c27](https://github.com/JEFFTheDev/pact-go/commit/0975c27) - fix: Dockerfile to reduce vulnerabilities (snyk-bot, Thu Mar 25 06:15:10 2021 +0000)
  * [f10eb78](https://github.com/JEFFTheDev/pact-go/commit/f10eb78) - build(deps): bump github.com/stretchr/testify from 1.4.0 to 1.7.0 (dependabot-preview[bot], Wed Jan 13 19:43:03 2021 +0000)

### v1.6.4 (09 August 2021)
  * [936315b](https://github.com/JEFFTheDev/pact-go/commit/936315b) - chore: tidy up consumer version selectors (Matt Fellows, Mon Aug 9 17:14:57 2021 +1000)
  * [01b8e81](https://github.com/JEFFTheDev/pact-go/commit/01b8e81) - fix: fixed linter errors in dsl/pact (antklim, Fri Jul 30 17:27:18 2021 +1000)
  * [5a5ef03](https://github.com/JEFFTheDev/pact-go/commit/5a5ef03) - fix: fixed linter errors in client/service_manager (antklim, Fri Jul 30 17:17:37 2021 +1000)
  * [72a05bf](https://github.com/JEFFTheDev/pact-go/commit/72a05bf) - fix: fixing linter errors in client/service_manager (antklim, Fri Jul 30 17:14:16 2021 +1000)
  * [60ddda1](https://github.com/JEFFTheDev/pact-go/commit/60ddda1) - fix: fixed linter errors in dsl/client (antklim, Fri Jul 30 17:09:11 2021 +1000)
  * [d88f9c3](https://github.com/JEFFTheDev/pact-go/commit/d88f9c3) - fix: fixing linter errors in dsl/client (antklim, Fri Jul 30 17:06:45 2021 +1000)
  * [4f57f48](https://github.com/JEFFTheDev/pact-go/commit/4f57f48) - fix: fixing linter error in dsl/pact (antklim, Fri Jul 30 16:57:38 2021 +1000)
  * [6bc61f4](https://github.com/JEFFTheDev/pact-go/commit/6bc61f4) - fix: fixing linter in dsl and client packages (antklim, Fri Jul 30 12:56:10 2021 +1000)
  * [1ccc97e](https://github.com/JEFFTheDev/pact-go/commit/1ccc97e) - fix: fixed linter errors in test files (antklim, Fri Jul 30 12:09:44 2021 +1000)

### v1.6.3 (09 August 2021)
  * [530d6ce](https://github.com/JEFFTheDev/pact-go/commit/530d6ce) - chore: tidy up consumer version selectors (Matt Fellows, Mon Aug 9 17:14:57 2021 +1000)
  * [de0bb5e](https://github.com/JEFFTheDev/pact-go/commit/de0bb5e) - feat: support release and environment consumer version selectors (antklim, Thu Jul 29 17:57:40 2021 +1000)

### v1.5.3 (03 July 2021)
  * [5903b70](https://github.com/JEFFTheDev/pact-go/commit/5903b70) - fix: handle verification response correctly (bbmokus, Fri Jul 2 16:49:11 2021 +0200)
  * [d879a4b](https://github.com/JEFFTheDev/pact-go/commit/d879a4b) - fix(readme): update link to pact broker (Travis Dixon, Mon Jun 21 08:35:50 2021 +1000)
  * [2c7a939](https://github.com/JEFFTheDev/pact-go/commit/2c7a939) - docs: clarify how the verification runs if both PactURLs and BrokerURL are provided (Matt Fellows, Wed Jun 16 11:16:31 2021 +1000)
  * [65f4304](https://github.com/JEFFTheDev/pact-go/commit/65f4304) - Pass consumer version selectors in VerifyMessageProviderRaw (James Halsall, Tue Jun 15 09:50:46 2021 +0100)
  * [629d508](https://github.com/JEFFTheDev/pact-go/commit/629d508) - docs: add note about roadmap/rust shared core (Matt Fellows, Sun May 30 12:18:41 2021 +1000)
  * [a1d589a](https://github.com/JEFFTheDev/pact-go/commit/a1d589a) - chore: update slack links (Matt Fellows, Sat May 29 22:57:03 2021 +1000)

doc.go
/*
Pact Go enables consumer driven contract testing, providing a mock service and
DSL for the consumer project, and interaction playback and verification
for the service provider project.

Consumer Tests

Consumer side Pact testing is an isolated test that ensures a given component
is able to collaborate with another (remote) component. Pact will automatically
start a Mock server in the background that will act as the collaborators' test
double.

This implies that any interactions expected on the Mock server will be validated,
meaning a test will fail if all interactions were not completed, or if unexpected
interactions were found:

A typical consumer-side test would look something like this:

	func TestLogin(t *testing.T) {

		// Create Pact client
		pact := Pact{
			Consumer: "My Consumer",
			Provider: "My Provider",
		}
		// Shuts down Mock Service when done
		defer pact.Teardown()

		// Pass in your test case as a function to Verify()
		var test = func() error {
			_, err := http.Get("http://localhost:8000/")
			return err
		}

    // Set up our expected interactions.
    pact.
      AddInteraction().
      Given("User foo exists").
      UponReceiving("A request to get foo").
      WithRequest(dsl.Request{
        Method:  "GET",
        Path:    dsl.String("/foobar"),
        Headers: dsl.MapMatcher{"Content-Type": "application/json"},
      }).
      WillRespondWith(dsl.Response{
        Status:  200,
        Headers: dsl.MapMatcher{"Content-Type": "application/json"},
        Body:    dsl.Match(&Foo{})
      })

    // Verify
    if err := pact.Verify(test); err != nil {
      log.Fatalf("Error on Verify: %v", err)
    }
  }

If this test completed successfully, a Pact file should have been written to
./pacts/my_consumer-my_provider.json containing all of the interactions
expected to occur between the Consumer and Provider.

Matching

In addition to verbatim value matching, you have 3 useful matching functions
in the `dsl` package that can increase expressiveness and reduce brittle test
cases.

	Term(example, matcher)	tells Pact that the value should match using a given regular expression, using `example` in mock responses. `example` must be a string.
	Like(content)		tells Pact that the value itself is not important, as long as the element _type_ (valid JSON number, string, object etc.) itself matches.
	EachLike(content, min)	tells Pact that the value should be an array type, consisting of elements like those passed in. `min` must be >= 1. `content` may be a valid JSON value: e.g. strings, numbers and objects.

Here is a complex example that shows how all 3 terms can be used together:

  	body :=
		Like(map[string]interface{}{
			"response": map[string]interface{}{
				"name": Like("Billy"),
        "type": Term("admin", "admin|user|guest"),
        "items": EachLike("cat", 2)
			},
		})

This example will result in a response body from the mock server that looks like:

  {
    "response": {
      "name": "Billy",
      "type": "admin",
      "items": [
        "cat",
        "cat"
      ]
    }
  }

See the examples in the dsl package and the matcher tests
(https://github.com/JEFFTheDev/pact-go/blob/master/dsl/matcher_test.go)
for more matching examples.

NOTE: You will need to use valid Ruby regular expressions
(http://ruby-doc.org/core-2.1.5/Regexp.html) and double escape backslashes.

Read more about flexible matching (https://github.com/JEFFTheDev/pact-ruby/wiki/Regular-expressions-and-type-matching-with-Pact.

Provider Tests

Provider side Pact testing, involves verifying that the contract - the Pact file
- can be satisfied by the Provider.

A typical Provider side test would like something like:

	func TestProvider_PactContract(t *testing.T) {
	  // Create Pact
		pact := Pact{}
		go startMyAPI("http://localhost:8000")

		pact.VerifyProvider(types.VerifyRequest{
			ProviderBaseURL:        "http://localhost:8000",
			PactURLs:               []string{"./pacts/my_consumer-my_provider.json"},
			ProviderStatesSetupURL: "http://localhost:8000/setup",
		})
	}

The `VerifyProvider` will handle all verifications, treating them as subtests
and giving you granular test reporting. If you don't like this behaviour, you may
call `VerifyProviderRaw` directly and handle the errors manually.

Note that `PactURLs` may be a list of local pact files or remote based
urls (possibly from a Pact Broker
- http://docs.pact.io/documentation/sharings_pacts.html).

Pact reads the specified pact files (from remote or local sources) and replays
the interactions against a running Provider. If all of the interactions are met
we can say that both sides of the contract are satisfied and the test passes.

Provider Verification

When validating a Provider, you have 3 options to provide the Pact files:

1. Use "PactURLs" to specify the exact set of pacts to be replayed:

	response, err = pact.VerifyProvider(types.VerifyRequest{
		ProviderBaseURL:        "http://myproviderhost",
		PactURLs:               []string{"http://broker/pacts/provider/them/consumer/me/latest/dev"},
		ProviderStatesSetupURL: "http://myproviderhost/setup",
		BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
		BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
		BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
	})

	2. Use "PactBroker" to automatically find all of the latest consumers:

	response, err = pact.VerifyProvider(types.VerifyRequest{
		ProviderBaseURL:        "http://myproviderhost",
		BrokerURL:              brokerHost,
		ProviderStatesSetupURL: "http://myproviderhost/setup",
		BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
		BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
		BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
	})

	3. Use "PactBroker" and "Tags" to automatically find all of the latest consumers:

	response, err = pact.VerifyProvider(types.VerifyRequest{
		ProviderBaseURL:        "http://myproviderhost",
		BrokerURL:              brokerHost,
		Tags:                   []string{"master", "sit4"},
		ProviderStatesSetupURL: "http://myproviderhost/setup",
		BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
		BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
		BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
	})

Options 2 and 3 are particularly useful when you want to validate that your
Provider is able to meet the contracts of what's in Production and also the latest
in development.

See this [article](http://rea.tech/enter-the-pact-matrix-or-how-to-decouple-the-release-cycles-of-your-microservices/)
for more on this strategy.

Provider States

Each interaction in a pact should be verified in isolation, with no context
maintained from the previous interactions. So how do you test a request that
requires data to exist on the provider? Provider states are how you achieve
this using Pact.

Provider states also allow the consumer to make the same request with different
expected responses (e.g. different response codes, or the same resource with a
different subset of data).

States are configured on the consumer side when you issue a dsl.Given() clause
with a corresponding request/response pair.

Configuring the provider is a little more involved, and (currently) requires
running an API endpoint to configure any [provider states](http://docs.pact.io/documentation/provider_states.html) during the
verification process. The option you must provide to the dsl.VerifyRequest
is:

  ProviderStatesSetupURL: 	POST URL to set the provider state (see types.ProviderState)

An example route using the standard Go http package might look like this:

  // Handle a request from the verifier to configure a provider state (ProviderStatesSetupURL)
  mux.HandleFunc("/setup", func(w http.ResponseWriter, req *http.Request) {
    w.Header().Add("Content-Type", "application/json")

    // Retrieve the Provider State
    var state types.ProviderState

    body, _ := ioutil.ReadAll(req.Body)
    req.Body.Close()
    json.Unmarshal(body, &state)

    // Configure database for different states
    if state.State == "User A exists" {
      svc.userDatabase = aExists
    } else if state.State == "User A is unauthorized" {
      svc.userDatabase = aUnauthorized
    } else {
      svc.userDatabase = aDoesNotExist
    }
  })

See the examples or read more at http://docs.pact.io/documentation/provider_states.html.

Publishing Pacts to a Broker and Tagging Pacts

See the Pact Broker (http://docs.pact.io/documentation/sharings_pacts.html)
documentation for more details on the Broker and this article
(http://rea.tech/enter-the-pact-matrix-or-how-to-decouple-the-release-cycles-of-your-microservices/)
on how to make it work for you.

Publishing using Go code:

	pact.PublishPacts(types.PublishRequest{
		PactBroker:             "http://pactbroker:8000",
		PactURLs:               []string{"./pacts/my_consumer-my_provider.json"},
		ConsumerVersion:        "1.0.0",
		Tags:                   []string{"master", "dev"},
	})

Publishing from the CLI:

Use a cURL request like the following to PUT the pact to the right location,
specifying your consumer name, provider name and consumer version.

	curl -v -XPUT \-H "Content-Type: application/json" \
	-d@spec/pacts/a_consumer-a_provider.json \
	http://your-pact-broker/pacts/provider/A%20Provider/consumer/A%20Consumer/version/1.0.0

Using the Pact Broker with Basic authentication

The following flags are required to use basic authentication when
publishing or retrieving Pact files to/from a Pact Broker:

	BrokerUsername	uername for Pact Broker basic authentication
	BrokerPassword	password for Pact Broker basic authentication

Output Logging

Pact Go uses a simple log utility (logutils - https://github.com/hashicorp/logutils)
to filter log messages. The CLI already contains flags to manage this,
should you want to control log level in your tests, you can set it like so:

	pact := Pact{
	  ...
		LogLevel: "DEBUG", // One of DEBUG, INFO, ERROR, NONE
	}
*/
package main
